protocol/xmrmaker/instance.go

```diff
@@ -198,6 +188,53 @@ func (inst *Instance) createOngoingSwap(s *swap.Info) error {
 	return nil
 }
 
+func (inst *Instance) completeSwap(s *swap.Info, skA *mcrypto.PrivateSpendKey) error {
+	// calc counterparty's swap private view key
+	vkA, err := skA.View()
+	if err != nil {
+		return err
+	}
+
+	// fetch our swap private spend key
+	skB, err := inst.backend.RecoveryDB().GetSwapPrivateKey(s.ID)
+	if err != nil {
+		return err
+	}
+
+	// fetch our swap private view key
+	vkB, err := skB.View()
+	if err != nil {
+		return err
+	}
+
+	kpAB := pcommon.GetClaimKeypair(
+		skA, skB,
+		vkA, vkB,
+	)
+
+	// we save the counterparty's public keys in case they send public keys derived
+	// in a non-standard way.
+	xmrtakerPublicKeys, err := inst.backend.RecoveryDB().GetXMRTakerSwapKeys(s.ID)
+	if err != nil {
+		return err
+	}
+
+	address := mcrypto.SumSpendAndViewKeys(
+		xmrtakerPublicKeys, kpAB.PublicKeyPair()).Address(inst.backend.Env())
```

Comment:
I assume this address is the same as `kpAB.Address(inst.backend.Env())`.  If so, you can skip passing it down 2 more levels of code and just call `CreateSpendWalletFromKeys` instead of the new function `CreateSpendWalletFromKeysAndAddress`.

Reply:
it's not necessarily the same address, which is why I introduced the new function. the xmrtaker can send us public view keys that aren't derived the usual way, and there's no way to check if it is or not (unless we add some complex zk proof or something), in which case the address we lock the xmr in will not be the same as when we get the address from the summed private keys.

Reply:
the other solution is to get the xmrtaker to also send the xmrmaker its private view key, but I'd rather not make unnecessary protocol changes. it would also fix this issue though

Reply:
Isn't this code path only creating spend wallets? For spend wallets, the address is always computed directly from the private keys.  For ViewOnly wallets, I created `CreateViewOnlyWalletFromKeys`, which does take the address as a parameter.

Reply:
yes it's creating the spend wallet, but the address still has to match that in which we locked the XMR in

Reply:
Do we check that the address matches before doing the expensive operation of creating a new `monero-wallet-rpc` instance? On a relative scale, computing the address very cheap. As long as it is not in a loop, invoking the address function multiple times is not an issue.

`CreateSpendWalletFromKeys` and the new, `CreateSpendWalletFromKeysAndAddress`, are the exact same function. Let's only keep one of them.  I liked the original, because I thought it made the code more self documenting. It lets the reader better understand the difference between spend and view-only wallets at creation time.  That with spend wallets, the address is a derived value, while with view-only wallets, it is a mandatory, non-derived value critical to create the wallet. When creating a spend wallet, it is very well vetted at this point that the address will always match that which we calculate from the private keys. Our unit tests for this have never failed.

Reply:
@dimalinux the PR has been redone, pls re-review